Zotero/Controllers/IdleTimerController.swift

```diff
@@ -8,35 +8,96 @@
+            activeTimer.suspend()
+            activeTimer.schedule(deadline: .now() + DispatchTimeInterval.seconds(Self.customIdleTimerTimemout))
+            activeTimer.resume()
+        }
+    }
 
-    func disable() {
+    func startCustomIdleTimer() {
+        guard !ProcessInfo.processInfo.isLowPowerModeEnabled else { return }
         inMainThread { [weak self] in
-            guard let self = self else { return }
-            self.activeProcesses += 1
-            UIApplication.shared.isIdleTimerDisabled = true
-            DDLogInfo("IdleTimerController: disable idle timer \(self.activeProcesses)")
+            guard let self else { return }
+            activeProcesses += 1
+            DDLogInfo("IdleTimerController: disable idle timer \(activeProcesses)")
+            guard activeTimer == nil else { return }
+            set(disabled: true)
+            startTimer(controller: self)
+        }
+
+        func startTimer(controller: IdleTimerController) {
+            let timer = DispatchSource.makeTimerSource(flags: [], queue: .main)
+            timer.schedule(deadline: .now() + DispatchTimeInterval.seconds(Self.customIdleTimerTimemout))
+            timer.setEventHandler(handler: { [weak controller] in
+                controller?.forceStopIdleTimer()
+            })
+            timer.resume()
+            controller.activeTimer = timer
         }
     }
 
-    func enable() {
+    func stopCustomIdleTimer() {
         inMainThread { [weak self] in
-            guard let self = self else { return }
-            guard self.activeProcesses > 0 else {
+            guard let self, activeProcesses > 0 else {
                 DDLogWarn("IdleTimerController: tried to enable idle timer with no active processes")
                 return
             }
+            activeProcesses -= 1
 
-            self.activeProcesses -= 1
-
-            DDLogInfo("IdleTimerController: enable idle timer \(self.activeProcesses)")
+            DDLogInfo("IdleTimerController: enable idle timer \(activeProcesses)")
 
-            guard self.activeProcesses == 0 else { return }
-            UIApplication.shared.isIdleTimerDisabled = false
+            guard activeProcesses == 0 else { return }
+            set(disabled: false)
+            activeTimer?.suspend()
+            activeTimer?.setEventHandler(handler: nil)
+            activeTimer?.cancel()
+            /*
+             If the timer is suspended, calling cancel without resuming
+             triggers a crash. This is documented here https://forums.developer.apple.com/thread/15902
+             */
+            activeTimer?.resume()
+            activeTimer = nil
         }
     }
+
+    private func set(disabled: Bool) {
+        UIApplication.shared.isIdleTimerDisabled = disabled
+    }
+
+    private func forceStopIdleTimer() {
+        DDLogInfo("IdleTimerController: force stop timer")
+        activeProcesses = 0
+        stopCustomIdleTimer()
```

Comment:
This call will fail in the first guard of `stopCustomIdleTimer`

Reply:
Fixed in b81703099fea811fc4dda794b99a9b6c4a36cc17